app/contracts/projects/archive_contract.rb

```diff
@@ -27,16 +27,44 @@
 #++
 
 module Projects
-  class ArchiveContract < ModelContract
-    include RequiresAdminGuard
-    include Projects::Archiver
-
+  class ArchiveContract < ::BaseContract
     validate :validate_no_foreign_wp_references
+    validate :validate_has_archive_project_permission
 
     protected
 
-    def validate_model?
-      false
+    # Check that there is no wp of a non descendant project that is assigned
+    # to one of the project or descendant versions
+    def validate_no_foreign_wp_references
+      version_ids = model.rolled_up_versions.select(:id)
+
+      exists = WorkPackage
+                 .where.not(project_id: model.self_and_descendants.select(:id))
+                 .exists?(version_id: version_ids)
+
+      errors.add :base, :foreign_wps_reference_version if exists
+    end
+
+    def validate_has_archive_project_permission
+      validate_can_archive_project
+      validate_can_archive_subprojects
+    end
+
+    def validate_can_archive_project
+      return if user.allowed_to?(:archive_project, model)
+
+      errors.add :base, :error_unauthorized
+    end
+
+    def validate_can_archive_subprojects
+      return if errors.any?
+
+      subprojects_with_missing_permission = model.descendants.reject do |subproject|
+        user.allowed_to?(:archive_project, subproject)
+      end
+      if subprojects_with_missing_permission.any?
+        errors.add :base, :archive_permission_missing_on_subprojects
+      end
```

Comment:
Why not combine the two loops?
```
unless model.descendants.all? { |subproject| user.allowed_to?(:archive_project, subproject) }
  errors.add :base, :archive_permission_missing_on_subprojects  
end
```

Reply:
Because it was easier to read it on multiple lines and without any `unless`. When I see `unless`, my brain has to convert it to `if not`.

On the other hand, the version you propose is more performant as it will stop on the first subproject where the user does not have the permission, whereas my version will scan for all subprojects.

Maybe I'd like to write this:

```
if model.descendants.any? { |subproject| user.forbidden_to?(:archive_project, subproject) }
  errors.add :base, :archive_permission_missing_on_subprojects  
end
```

Can I add a `forbidden_to?` next to the `allowed_to?`?

And after digging deeper, it appears that `user#allowed_to?` accepts an array of projects, so it can be written like this instead:

```
unless user.allowed_to?(:archive_project, model.descendants)
  errors.add :base, :archive_permission_missing_on_subprojects  
end
```

Seems better.

Reply:
I changed it to an early return:

```
    def validate_can_archive_subprojects
      return if errors.any?
      return if user.allowed_to?(:archive_project, model.descendants)

      errors.add :base, :archive_permission_missing_on_subprojects
    end
```

Previous version would have been helpful if the subprojects names were added to the error message, but that's not the case.

Reply:
But it does not work because `user.allowed_to?(action, projects)` returns `false` if `projects` is an empty array.

I'll look into it tomorrow.

Reply:
Done in d91666d